mlflow/utils/proto_json_utils.py

```diff
@@ -370,36 +397,54 @@ def parse_tf_serving_input(inp_dict, schema=None):
     import numpy as np
 
     def cast_schema_type(input_data):
+        types_dict = {}
         if schema is not None:
             if schema.has_input_names():
-                input_names = schema.input_names()
-                if len(input_names) == 1 and isinstance(input_data, list):
-                    # for schemas with a single column, match input with column
-                    input_data = {input_names[0]: input_data}
-                if not isinstance(input_data, dict):
-                    raise MlflowException(
-                        "Failed to parse input data. This model contains a tensor-based model"
-                        " signature with input names, which suggests a dictionary input mapping"
-                        f" input name to tensor, but an input of type {type(input_data)} was found."
-                    )
-                type_dict = dict(zip(schema.input_names(), schema.numpy_types()))
-                for col_name in input_data.keys():
-                    input_data[col_name] = np.array(
-                        input_data[col_name], dtype=type_dict.get(col_name)
-                    )
+                types_dict = schema.input_dict()
+                if (
+                    len(types_dict) == 1
+                    and isinstance(input_data, list)
+                    and not any(isinstance(x, dict) for x in input_data)
+                ):
+                    # for schemas with a single column (not List[Dict]), match input with column
+                    input_data = {list(types_dict.keys())[0]: input_data}
+            # Un-named schema should only contain a single column
+            elif not isinstance(input_data, list):
+                raise MlflowException(
+                    "Failed to parse input data. This model contains an un-named tensor-based"
+                    " model signature which expects a single n-dimensional array as input,"
+                    f" however, an input of type {type(input_data)} was found."
+                )
+        if isinstance(input_data, dict):
+            input_data = {
+                col: convert_data_type(data, types_dict.get(col))
```

Comment:
is `types_dict` populated if `schema` is None? 
Can we validate that condition in tests?

Reply:
The previous existing logic simply ignored validation if the schema is undefined. We should verify that we're preserving this behavior and only doing the conversion validation iff we have a schema populated

Reply:
`types_dict` is not populated if schema is None, and in that case `convert_data_type` should just be `np.array(data)`.

Reply:
ahhhhh sorry I missed the first declaration in that function. If there's no spec, use inferred type casting by numpy. SGTM!